lib/Business/Tax/VAT/Validation.pm

```diff
@@ -360,42 +360,53 @@ sub _is_res_ok {
     $self->{informations}={};
     $res=~s/[\r\n]/ /g;
     $self->{response} = $res;
-    if ($code == 200) {
-        if ($res=~m/<valid> *(.*?) *<\/valid>/) {
-            my $v = $1;
-            if ($v eq 'true' || $v eq '1') {
-                if ($res=~m/<name> *(.*?) *<\/name>/) {
-                    $self->{informations}{name} = $1
-                }
-                if ($res=~m/<address> *(.*?) *<\/address>/) {
-                    $self->{informations}{address} = $1
-                }
-                $self->_set_error( -1, 'Valid VAT Number');
-                return 1;
-            } else {
-                return $self->_set_error( 2, 'Invalid VAT Number ('.$v.')');
-            }
+
+    # Check for a SOAP fault first (sometimes we can get one even with a
+    # successful HTTP status)
+    if ($res=~m/<faultcode> *(.*?) *<\/faultcode> *<faultstring> *(.*?) *<\/faultstring>/) {
+        my $faultcode   = $1;
+        my $faultstring = $2;
```

Comment:
You're parsing XML with regex. [Do not parse XML with regex](https://stackoverflow.com/a/1732454).

Also: are named captures too recent a feature for a CPAN module, or could you add named captures and a /x switch and make this regex slightly more readable and robust?

Failing that, at least use a delimeter other than / so you don't have to backslash the /s in XML tags.

Reply:
I was looking more for feedback on the changes I'm making for this bugfix, not the existing code I've moved around a bit, and which, yes, needs some serious improvement at some point - but I haven't time for that right now!